src/main/java/org/terasology/joshariasSurvival/systems/SheepShearingSystem.java

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.joshariasSurvival.systems;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.engine.core.Time;
+import org.terasology.engine.entitySystem.entity.EntityManager;
+import org.terasology.engine.entitySystem.entity.EntityRef;
+import org.terasology.engine.entitySystem.event.ReceiveEvent;
+import org.terasology.engine.entitySystem.systems.BaseComponentSystem;
+import org.terasology.engine.entitySystem.systems.RegisterSystem;
+import org.terasology.engine.logic.characters.events.AttackEvent;
+import org.terasology.wildAnimals.event.AnimalSpawnEvent;
+import org.terasology.engine.logic.delay.DelayManager;
+import org.terasology.engine.logic.delay.PeriodicActionTriggeredEvent;
+import org.terasology.engine.logic.inventory.ItemComponent;
+import org.terasology.engine.registry.In;
+import org.terasology.joshariasSurvival.components.ShearableSheepComponent;
+
+
+import java.util.ArrayList;
+
+@RegisterSystem
+public class SheepShearingSystem extends BaseComponentSystem {
+    public static final int HITS_TO_SHEAR = 5;
+    public static final ArrayList<ItemComponent> SHEARING_ITEMS = null; //TODO yet to decide
+    public static final int HAIR_REGROWTH_TIME = 3*60*1000; // In ms
+    public static final String HAIR_REGROWTH_ACTION_ID = "Check hair regrowth";
+    private static final Logger logger = LoggerFactory.getLogger(SheepShearingSystem.class);
+
+    @In
+    protected Time time;
+
+    @In
+    private DelayManager delayManager;
+
+    @In
+    private EntityManager entityManager;
+
+    @ReceiveEvent
+    public void onAnimalSpawn(AnimalSpawnEvent event, EntityRef entityRef) {
+        if(entityRef.getParentPrefab().toString().contains("WildAnimals:sheep")) {
+            entityRef.addComponent(new ShearableSheepComponent());
+        }
+    }
+
+    @ReceiveEvent(components = {ShearableSheepComponent.class})
+    public void onAnimalSpawn(AttackEvent event, EntityRef entityRef) {
+        ShearableSheepComponent component = entityRef.getComponent(ShearableSheepComponent.class);
+        if(!component.isSheared()) {
+            int clicks = component.getClicks();
+            if (clicks < HITS_TO_SHEAR) {
+                component.setClicks(++clicks);
+            } else {
+                component.setClicks(0);
+                component.setSheared(true);
+                component.setLastShearingTimestamp(time.getGameTimeInMs());
+                delayManager.addPeriodicAction(entityRef, HAIR_REGROWTH_ACTION_ID, 0, 10 * 1000);
```

Comment:
suggestion: Since we already know `HAIR_REGROWTH_TIME` when we go to set up this delayed action, we needn't set up a periodic action with an arbitrary time of ten seconds and then compare. You can do `addDelayedAction(entityref, ACTION_ID, HAIR_REGROWTH_TIME)`.

Reply:
Agreed, this does not need to be a periodic action. It's sufficient to schedule a delayed action in the event of shearing and, in case we expect sheep to spawn in unsheared state (do we?), on spawn.

Reply:
The motive behind using this is that if a chunk containing the sheep entity gets unloaded and loaded again, the hair regrowth should still work. For eg, a sheep gets sheared and then the chunk containing sheep gets unloaded and then the Delayed Action occurs and then the chunk gets loaded. IIRC @skaldarnar pointed this out while reviewing my proposal.

Reply:
Makes sense regarding the unsheared state.
When a formerly unloaded chunk gets loaded again, do we get another spawn notification or so? :thinking: tbh, I'm not entirely sure how this works... maybe @skaldarnar can shed some light?

Reply:
That is why we save the last shearing timestamp also.

Reply:
That is why we save the last shearing timestamp also.